### PR TITLE
New addon: Search sprites addon

### PR DESCRIPTION
**Pull request by BarelySmooth**
Resolves https://github.com/ScratchAddons/ScratchAddons/issues/3062

### Changes
Adds an addon to search sprites in the project editor.

Image showing the search sprites addon

### Reason for changes
21 people liked the idea.

### Tests
Tested on Edge and Firefox

This is the first time I've tried creating an addon. I might have done a lot of stuff wrongly, sorry for that.

### CODE REVIEW

**Comment by Samq64** [comment]
This works quite well, but it's not completely compatible with sprite folders. That's okay though because I can still search for closed folders and open them and it might actually be better that way since it won't keep opening closed folders.

File: addons/search-sprites/search-bar.css [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    background-color: #eaf1fb;
+    border: 0px;
+    border-bottom: 1px solid hsla(0, 0%, 1%, 0.15);
+    outline: none;
+    padding: 5px 10px;
+}
+
+input[type="search"]::-webkit-search-cancel-button,
+input[type="search"]::-webkit-search-decoration {
+    -webkit-appearance: auto;
```
Comment (RedGuy12) [inline addons/search-sprites/search-bar.css:14]: This only works in chromium.

File: addons/search-sprites/search-bar.css [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#sa-sprite-search-box {
+    color: black;
```
Comment (RedGuy12) [inline addons/search-sprites/search-bar.css:2]: Are there any classes built in to scratch that can style this?

File: addons/search-sprites/userscript.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+export default async function({ addon, global, console }) {
+    let sprites_container = document.getElementsByClassName('sprite-selector_items-wrapper_4bcOj')[0];
```
Comment (RedGuy12) [inline addons/search-sprites/userscript.js:2]: The ending hash will change. Use `[class*="..."]` with querySelector

File: addons/search-sprites/userscript.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    let search_box = document.createElement('input');
+    search_box.setAttribute('type', 'search');
+    search_box.setAttribute('placeholder', 'Search sprites...');
+    search_box.setAttribute('id', 'sa-sprite-search-box');
+    search_box.setAttribute('autocomplete', 'off');
+
+    await addon.tab.waitForElement("div.sprite-selector_items-wrapper_4bcOj");
```
Comment (RedGuy12) [inline addons/search-sprites/userscript.js:9]: Remove hash

File: addons/search-sprites/userscript.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    let sprite_selector_container = document.getElementsByClassName('sprite-selector_scroll-wrapper_3NNnc')[0];
+    sprite_selector_container.insertBefore(search_box, document.getElementsByClassName('sprite-selector_items-wrapper_4bcOj')[0]);
```
Comment (RedGuy12) [inline addons/search-sprites/userscript.js:12]: Hash

File: addons/search-sprites/userscript.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+
+    addon.tab.displayNoneWhileDisabled(search_box, { display: "block" });
+    addon.self.addEventListener("disabled", () => {
+        for (let i = 0; i < sprites_container.children.length; i++) {sprites_container.children[i].style.display = 'block'; console.log('disabled');}
```
Comment (RedGuy12) [inline addons/search-sprites/userscript.js:17]: Remove log

File: addons/search-sprites/userscript.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        } else {
+            for (let i = 0; i < sprites_container.children.length; i++) {
+                sprites_container.children[i].style.display = 'block';
+            }
+        }
+    }
+
+    
+    while (true) {
+        await addon.tab.waitForElement("div.sprite-selector_items-wrapper_4bcOj", {
```
Comment (RedGuy12) [inline addons/search-sprites/userscript.js:42]: Hash

File: addons/search-sprites/userscript.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+          markAsSeen: true,
+          reduxCondition: (state) => !state.scratchGui.mode.isPlayerOnly,
+        });
+
+        sprites_container = document.getElementsByClassName('sprite-selector_items-wrapper_4bcOj')[0];
+        sprite_selector_container = document.getElementsByClassName('sprite-selector_scroll-wrapper_3NNnc')[0];
+        sprite_selector_container.insertBefore(search_box, document.getElementsByClassName('sprite-selector_items-wrapper_4bcOj')[0]);
+    }
```
Comment (RedGuy12) [inline addons/search-sprites/userscript.js:50]: Hash

**Review by RedGuy12 — CHANGES_REQUESTED** [review]

**Comment by RedGuy12** [comment]
Also please enable actions here: https://github.com/BarelySmooth/ScratchAddons/actions

File: addons/search-sprites/search-bar.css (hunk 1 above)
Comment (BarelySmooth) [reply addons/search-sprites/search-bar.css]: That exists as a flag in Firefox. `layout.forms.input-type-search.enabled` is the name of the flag.
firefox

File: addons/search-sprites/search-bar.css (hunk 2 above)
Comment (BarelySmooth) [reply addons/search-sprites/search-bar.css]: @RedGuy12 just the color or the entire search box?

Comment (RedGuy12) [reply addons/search-sprites/search-bar.css]: no i dont mean remove `color: black` i mean use a different class to style everything

Comment (BarelySmooth) [reply addons/search-sprites/search-bar.css]: I'm not sure if such a class exists. I'll try checking.

(also, removing `color: black` was intentional)

File: addons/search-sprites/userscript.js [hunk 9]
```diff
@@ -1,4 +1,5 @@
 export default async function ({ addon, global, console }) {
+  console.log("hi");
```
Comment (RedGuy12) [inline addons/search-sprites/userscript.js:2]: cruft

File: addons/search-sprites/search-bar.css (hunk 1 above)
Comment (RedGuy12) [reply addons/search-sprites/search-bar.css]: we don't want them to have to use a flag

File: addons/search-sprites/search-bar.css (hunk 2 above)
Comment (RedGuy12) [reply addons/search-sprites/search-bar.css]: also im not sure if the current style completely fits in with scratch

Comment (BarelySmooth) [reply addons/search-sprites/search-bar.css]: but if it did, it would take up too much height

Comment (BarelySmooth) [reply addons/search-sprites/search-bar.css]: like this:
Screenshot 2021-12-29 at 10 37 10 AM

**Comment by hello-smile6** [comment]
> Resolves #3062
> 
> ### Changes
> Adds an addon to search sprites in the project editor.
> 
> Image showing the search sprites addon
> 
> ### Reason for changes
> 21 people liked the idea.
> 
> ### Tests
> Tested on Edge and Firefox
> 
> This is the first time I've tried creating an addon. I might have done a lot of stuff wrongly, sorry for that.

Nice!

File: addons/search-sprites/userscript.js (hunk 9 above)
Comment (hello-smile6) [reply addons/search-sprites/userscript.js]: I do that too

**Review by hello-smile6 — APPROVED** [review]
Good!

File: addons/search-sprites/search-bar.css (hunk 2 above)
Comment (hello-smile6) [reply addons/search-sprites/search-bar.css]: > like this: Screenshot 2021-12-29 at 10 37 10 AM

A bit too big

Comment (TheColaber) [reply addons/search-sprites/search-bar.css]: You can probably use the input_input-form_l9eYg class.

**Comment by GarboMuffin** [comment]
Merge https://github.com/BarelySmooth/ScratchAddons/pull/1 and I will be reasonably happy with the state of the addon code. It was mostly just refactoring. Should work the same. The style of the input is not something that I touched or care much about, but should probably be considered more if this eventually is to be merged

**Pull request merged**